xtm/xtm_api.cc

```diff
@@ -0,0 +1,249 @@
+	struct xtm_scsp_queue<union xtm_msg> queue;
+};
+
+static inline int
+xtm_queue_flush_pipe(struct xtm_queue *xtm_queue)
+{
+	unsigned char tmp[XTM_PIPE_SIZE];
+	ssize_t read_bytes;
+	while ((read_bytes = read(xtm_queue->readfd, tmp, sizeof(tmp))) < 0 &&
+	        errno == EINTR)
+		;
+	if (read_bytes <= 0 && errno != EAGAIN && errno != EWOULDBLOCK)
+		return -1;
+	return 0;
+}
+
+
+struct xtm_queue *
+xtm_queue_new(unsigned size)
+{
+	int filedes[2];
+	struct xtm_queue *queue = (struct xtm_queue *)
+		malloc(sizeof(struct xtm_queue) +
+		       size * sizeof(union xtm_msg));
+	if (queue == NULL)
+		return NULL;
+
+#ifdef TARANTOOL_XTM_USE_EVENTFD
+	if ((filedes[0] = eventfd(0, 0)) < 0)
+#else
+	if (pipe(filedes) < 0)
+#endif
+		goto free_queue;
+
+#ifdef TARANTOOL_XTM_USE_EVENTFD
+	queue->readfd = queue->writefd = filedes[0];
+#else
+	queue->readfd = filedes[0];
+	queue->writefd = filedes[1];
+#endif
+
+	if (fcntl(queue->readfd, F_SETFL, O_NONBLOCK) < 0 ||
+	    fcntl(queue->writefd, F_SETFL, O_NONBLOCK) < 0)
+		goto close_fd;
+	if (queue->queue.xtm_scsp_queue_init(size) < 0) {
+		errno = EINVAL;
+		goto close_fd;
+	}
+	return queue;
+
+close_fd:
+	close(queue->readfd);
+	if (queue->readfd != queue->writefd)
+		close(queue->writefd);
+free_queue:
+	free(queue);
+	return NULL;
+};
+
+int
+xtm_queue_delete(struct xtm_queue *queue)
+{
+	int rc = 0;
+	if (close(queue->readfd) < 0)
+		rc = -1;
+	if (queue->readfd != queue->writefd && close(queue->writefd) < 0)
+		rc = -1;
+	free(queue);
+	return rc;
+}
+
+int
+xtm_msg_probe(struct xtm_queue *queue)
+{
+	if (queue->queue.xtm_scsp_queue_free_count() == 0) {
+		errno = ENOBUFS;
+		return -1;
+	}
+	return 0;
+}
+
+unsigned
+xtm_msg_count(struct xtm_queue *queue)
+{
+	return queue->queue.xtm_scsp_queue_count();
+}
+
+int
+xtm_msg_notify(struct xtm_queue *queue)
+{
+	uint64_t tmp = 1;
+	ssize_t cnt;
+	/*
+	 * We must write 8 byte value, because for linux we
+	 * used eventfd, which require to write >= 8 byte at once.
+	 * Also in case of EINTR we retry to write
+	 */
+	while ((cnt = write(queue->writefd, &tmp, sizeof(tmp))) < 0 &&
+	       errno == EINTR)
+		;
+	return (cnt == sizeof(tmp) ? 0 : -1);
+}
+
+int
+xtm_fun_dispatch(struct xtm_queue *queue, fun_for_dispatch fun,
+		 void *fun_arg, int delayed)
+{
+	union xtm_msg xtm_msg;
+	xtm_msg.fun = fun;
+	xtm_msg.fun_arg = fun_arg;
+	if (queue->queue.xtm_scsp_queue_put(&xtm_msg, 1) != 1) {
+		errno = ENOBUFS;
+		return -1;
+	}
+	if (delayed == 0)
+		return xtm_msg_notify(queue);
+	return 0;
+}
+
+int
+xtm_fd(struct xtm_queue *queue)
+{
+	return queue->readfd;
+}
+
+int
+xtm_fun_invoke(struct xtm_queue *queue, int flushed)
+{
+	if (flushed != 0 && xtm_queue_flush_pipe(queue) < 0)
+		return -1;
+	struct xtm_scsp_queue_read_iterator<xtm_msg> iter;
+	iter.xtm_scsp_queue_iterator_begin(&queue->queue);
+	const union xtm_msg *xtm_msg;
+	unsigned cnt = 0;
+
+	while((xtm_msg = iter.xtm_scsp_queue_iterator_read()) != nullptr) {
```

Comment:
AFAIS, you do not provide any protection against sending an 'msg' instead of 'fun', and hence it would crash here when I 'invoke' an 'msg'. But if it is not allowed, it could be banned, and you could make the queue elements just 8 bytes size for pure-msg queues. This is another reason why mixing functions and messages might be not a good idea. But I see the usecase in httpng, and don't know really.